livebooks/soundwave/soundwave.livemd

```diff
@@ -68,134 +56,96 @@ defmodule Visualizer do
 
   require Membrane.Logger
 
-  @window_size 1000
+  # The amount of points visible in the chart. The more points, the better chart resolution,
+  # but higher CPU consumption.
+  @visible_points 1000
 
-  # seconds
+  # Last n seconds of audio visible in the chart. Increasing the duration
+  # lowers the chart resolution, so you may want to increase @visible_points
+  # accordingly.
   @window_duration 3
 
-  # Hz
+  # Frequency of plot updates. Doesn't impact the chart resolution.
   @plot_update_frequency 50
 
-  @points_per_update @window_size / (@window_duration * @plot_update_frequency)
+  @points_per_update @visible_points / (@window_duration * @plot_update_frequency)
 
-  def_input_pad :input,
-    accepted_format: %RawAudio{},
-    flow_control: :auto
+  def_input_pad(:input, accepted_format: %RawAudio{})
 
   @impl true
   def handle_init(_ctx, _opts) do
-    {[],
-     %{
-       chart: nil,
-       initial_pts: nil,
-       bytes_per_sample: nil,
-       sample_rate: nil,
-       sample_format: nil,
-       channels: nil,
-       samples: []
-     }}
-  end
-
-  defguardp has_stream_format_arrived(ctx) when ctx.pads.input.stream_format != nil
-
-  @impl true
-  def handle_stream_format(:input, stream_format, ctx, state)
-      when not has_stream_format_arrived(ctx) do
-    {_sign, bits_per_sample, _endianness} =
-      RawAudio.SampleFormat.to_tuple(stream_format.sample_format)
-
-    chart = create_chart(stream_format)
-    Kino.render(chart)
-
-    {[],
-     %{
-       state
-       | sample_rate: stream_format.sample_rate,
-         sample_format: stream_format.sample_format,
-         channels: stream_format.channels,
-         bytes_per_sample: :erlang.round(bits_per_sample / 8),
-         chart: chart
-     }}
+    {[], %{chart: nil, pts: nil, initial_pts: nil, samples: []}}
   end
 
   @impl true
-  def handle_stream_format(:input, _stream_format, _ctx, state) do
-    Membrane.Logger.warning(":input stream format received once again, ignoring.")
-    {[], state}
+  def handle_setup(_ctx, state) do
+    {[], %{state | chart: render_chart()}}
   end
 
   @impl true
   def handle_buffer(:input, buffer, ctx, state) do
     state = if state.initial_pts == nil, do: %{state | initial_pts: buffer.pts}, else: state
+    state = if state.pts == nil, do: %{state | pts: buffer.pts}, else: state
+    stream_format = ctx.pads.input.stream_format
+    sample_size = RawAudio.sample_size(stream_format)
+    sample_max = RawAudio.sample_max(stream_format)
 
     samples =
-      for <<sample::binary-size(state.bytes_per_sample) <- buffer.payload>> do
-        RawAudio.sample_to_value(sample, ctx.pads.input.stream_format)
+      for <<sample::binary-size(sample_size) <- buffer.payload>> do
+        RawAudio.sample_to_value(sample, stream_format) / sample_max
```

Comment:
Shouldn't it be more like: 
```suggestion
        RawAudio.sample_to_value(sample, stream_format) / (sample_max-sample_min)
```?
(I am a little bit worried it won't get rescaled properly in case we have a signed number as a sample value)

Reply:
I think PortAudio returns signed numbers by default. `sample_max` is around `2^15` and `sample_min` is around `-(2^15)`, so by dividing by `sample_max` we're getting `[-1, 1]`. After the proposed change, I think we'd get `[-0.5, 0.5]` 🤔